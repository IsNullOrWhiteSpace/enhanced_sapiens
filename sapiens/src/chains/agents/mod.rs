/// OODA agents
pub mod ooda;

use crate::chains::Outcome;
use crate::context;
use crate::prompt::Task;
use crate::tools::ToolUseError;

/// Error from the agent
#[derive(thiserror::Error, Debug)]
pub enum Error {
    /// Failed to add to the chat history
    #[error("Failed to add to the chat history: {0}")]
    ChatHistoryError(#[from] context::Error),
    /// Error from the model
    #[error("Error from the model: {0}")]
    ModelError(#[from] crate::models::Error),
}

/// Format the outcome of a task
pub(crate) fn format_outcome(
    task: &Task,
    invocation_count: &usize,
    tool_name: &Option<String>,
    outcome: &Outcome,
) -> String {
    match outcome {
        Outcome::Success { result } => {
          