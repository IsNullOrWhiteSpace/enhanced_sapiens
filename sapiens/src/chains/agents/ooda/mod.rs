
/// OODA one-step agent
pub mod one_step;

/// Multistep OODA agents
///
/// - Observer: observe the environment
/// - Orienter: implicit guidance and control
/// - Decider: decision-making
/// - Actor: act on the environment
pub mod multistep;