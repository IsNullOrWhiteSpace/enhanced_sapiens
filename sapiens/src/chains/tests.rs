
use std::sync::Arc;

use indoc::indoc;
use serde_yaml::Value;
use tokio::sync::Mutex;

use super::*;
use crate::tools::{FieldFormat, Format, TerminalTool, Tool, ToolDescription, ToolUseError};
use crate::void_observer;

struct SimpleAgent {}

impl From<()> for Error {
    fn from(_value: ()) -> Self {
        panic!("This should not happen")
    }
}

#[async_trait::async_trait]
impl Agent for SimpleAgent {
    type Error = ();

    async fn act(&self, _context: &Context) -> Result<Message, ()> {
        Ok(Message::Observation {
            content: "Hello".to_string(),
            usage: None,
        })
    }
}

#[derive(Default)]
struct ConcludeTool {
    done: Mutex<Option<String>>,
}

#[async_trait::async_trait]
impl Tool for ConcludeTool {
    fn description(&self) -> ToolDescription {
        ToolDescription {
            name: "ConcludeTool".to_string(),
            description: "A tool to conclude a task".to_string(),
            parameters: Format {
                fields: vec![FieldFormat {
                    name: "conclusion".to_string(),
                    r#type: "str".to_string(),
                    optional: false,
                    description: "The conclusion".to_string(),
                }],
            },
            responses_content: Format::default(),
        }
    }

    async fn invoke(&self, _input: Value) -> Result<Value, ToolUseError> {
        let mut done = self.done.lock().await;

        if done.is_some() {
            return Err(ToolUseError::InvocationFailed(
                "This task is already done.".to_string(),
            ));
        }

        // set done
        *done = Some("Done".to_string());

        Ok(Value::Null)
    }
}

#[async_trait::async_trait]
impl TerminalTool for ConcludeTool {
    async fn is_done(&self) -> bool {
        let done = self.done.lock().await;
        done.is_some()
    }

    async fn take_done(&self) -> Option<TerminationMessage> {
        {
            let mut done = self.done.lock().await;
            done.take().map(|input| TerminationMessage {
                conclusion: input,
                original_question: "tbd".to_string(),
            })
        }
    }
}

#[tokio::test]
async fn observes_too_much() {
    let toolbox = {
        let mut toolbox = Toolbox::default();
        toolbox.add_terminal_tool(ConcludeTool::default()).await;
        toolbox
    };
    let observer = void_observer();
    let observer = Arc::downgrade(&observer);

    let scheduler = Box::new(schedulers::SingleAgentScheduler::new(
        10,
        Box::new(SimpleAgent {}),
        observer.clone(),
    ));
    let mut runtime = Runtime::new(toolbox, scheduler, observer).await.unwrap();

    let terminal_state = runtime.run().await;

    runtime
        .context
        .messages
        .iter()
        .for_each(|m| println!("{:?}", m));

    assert!(terminal_state.is_err());
}

struct NoAsSimpleAgent {}

#[async_trait::async_trait]
impl Agent for NoAsSimpleAgent {
    type Error = ();

    async fn act(&self, context: &Context) -> Result<Message, ()> {
        if context.messages.is_empty() {
            Ok(Message::Observation {
                content: "Hello".to_string(),
                usage: None,
            })
        } else {
            Ok(Message::Action {
                content: indoc! {r#"
                ```yaml
                tool_name: ConcludeTool
                parameters:
                    conclusion: "Done"
                ```
                "#
                }
                .to_string(),
                usage: None,
            })
        }
    }
}

#[tokio::test]
async fn observes_and_conclude() {
    let toolbox = {
        let mut toolbox = Toolbox::default();
        toolbox.add_terminal_tool(ConcludeTool::default()).await;
        toolbox
    };

    let observer = void_observer();
    let observer = Arc::downgrade(&observer);

    let scheduler = Box::new(schedulers::SingleAgentScheduler::new(
        10,
        Box::new(NoAsSimpleAgent {}),
        observer.clone(),
    ));
    let mut runtime = Runtime::new(toolbox, scheduler, observer).await.unwrap();

    let terminal_state = runtime.run().await;

    runtime
        .context
        .messages
        .iter()
        .for_each(|m| println!("{:?}", m));

    assert!(terminal_state.is_ok());

    let terminal_state = terminal_state.unwrap();
    assert_eq!(terminal_state.messages.len(), 1);

    let message = &terminal_state.messages[0];
    assert_eq!(message.conclusion, "Done");
}