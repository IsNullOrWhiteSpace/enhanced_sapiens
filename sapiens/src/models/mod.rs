pub mod openai;
pub mod vertex_ai;

use std::fmt::{Debug, Display};
use std::str::FromStr;
use std::sync::Arc;

use serde::{Deserialize, Serialize};

use crate::context::ChatEntry;

/// A model reference
pub type ModelRef = Arc<Box<dyn Model>>;

/// Errors from the models
#[derive(thiserror::Error, Debug)]
pub enum Error {
    /// The openai error
    #[error("Model invocation failed")]
    OpenAIError(#[from] openai::OpenAIError),
    /// No response from the model
    #[error("No response from the model")]
    NoResponseFromModel,
    /// The model is not supported
    #[error("Model not supported: {0}")]
    ModelNotSupported(String),
    /// Vertex AI error
    #[error("Vertex AI error: {0}")]
    VertexAIError(#[from] gcp_vertex_ai_generative_language::Error),
    /// Filtered output
    #[error("Filtered output")]
    Filtered,
}

/// Roles in the conversation
#[derive(Debug, Serialize, Deserialize, Clone, Default, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum Role {
    /// The system
    System,
    /// The user
    #[default]
    User,
    /// The assistant
    Assistant,
    /// Function call
    Function,
    /// Tool call
    Tool,
}

impl Display for Role {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Role::System => write!(f, "system"),
            Role::User => write!(f, "user"),
            Role::Assistant => write!(f, "assistant"),
            Role::Function => write!(f, "function"),
            Role::Tool => write!(f, "tool"),
        }
    }
}

// FUTURE(ssoudan) support pure