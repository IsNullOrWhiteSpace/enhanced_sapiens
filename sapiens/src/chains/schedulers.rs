
use super::*;
use crate::chains;

/// A simple scheduler that can be used to schedule agents
///
/// It has only one agent and a maximum number of steps.
/// Since it has only one agent, it will always schedule the same agent.
pub struct SingleAgentScheduler<E> {
    remaining_steps: usize,
    agent: Box<dyn Agent<Error = E>>,
    #[allow(dead_code)]
    observer: WeakRuntimeObserver,
}

impl<E> SingleAgentScheduler<E> {
    /// Create a new scheduler with a maximum number of steps and an agent
    pub fn new(
        max_steps: usize,
        agent: Box<dyn Agent<Error = E>>,
        observer: WeakRuntimeObserver,
    ) -> Self {
        Self {
            remaining_steps: max_steps,
            agent,
            observer,
        }
    }
}

#[async_trait::async_trait]
impl<E> Scheduler for SingleAgentScheduler<E>
where
    chains::Error: From<E>,
{
    async fn schedule(&mut self, context: &Context) -> Result<Message, Error> {
        if self.remaining_steps == 0 {
            return Err(Error::MaxStepsReached);
        }
        self.remaining_steps -= 1;

        let agent = self.agent.as_ref();

        let message = agent.act(context).await?;

        Ok(message)
    }
}

/// Scheduler that schedules multiple agents in a fixed order
pub struct MultiAgentScheduler<E> {
    remaining_steps: usize,
    next_agent: usize,
    agents: Vec<Box<dyn Agent<Error = E>>>,
    #[allow(dead_code)]
    observer: WeakRuntimeObserver,
}

impl<E> MultiAgentScheduler<E> {
    /// Create a new scheduler with a maximum number of steps and a list of
    /// agents
    pub fn new(
        max_steps: usize,
        agents: Vec<Box<dyn Agent<Error = E>>>,
        observer: WeakRuntimeObserver,
    ) -> Self {
        Self {
            remaining_steps: max_steps,
            next_agent: 0,
            agents,
            observer,
        }
    }
}

#[async_trait::async_trait]
impl<E> Scheduler for MultiAgentScheduler<E>
where
    chains::Error: From<E>,
{
    async fn schedule(&mut self, context: &Context) -> Result<Message, Error> {
        if self.remaining_steps == 0 {
            return Err(Error::MaxStepsReached);
        }
        self.remaining_steps -= 1;

        if self.next_agent >= self.agents.len() {
            self.next_agent = 0;
        }

        let agent = self.agents.get_mut(self.next_agent).unwrap();
        self.next_agent += 1;

        let message = agent.act(context).await?;

        Ok(message)
    }
}