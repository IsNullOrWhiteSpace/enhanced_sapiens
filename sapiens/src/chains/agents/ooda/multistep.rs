use std::fmt::{Debug, Formatter};

use tracing::{debug, trace};

use crate::chains::agents::{format_outcome, Error};
use crate::chains::{Context, Message};
use crate::context::{ChatEntry, ChatHistory};
use crate::models::Role;
use crate::tools::toolbox::Toolbox;
use crate::{chains, prompt, SapiensConfig, WeakRuntimeObserver};

const PREFIX: &str = r"You are part of a group of cooperating assistants named Sapiens. Use available tools to answer the question as best as you can.
You will collectively proceed iteratively using an OODA loop. Don't overstep your role.

- Action response will be provided. 
- Never produce the response of an Action. 
- Only use YAML for the Action.
- The loop will repeated until you have the answer to the original question. 
- No task is complete until the Conclude Tool is used to provide the answer. 
";

const TOOL_PREFIX: &str = r"
# The following are the ONLY Tools one can use for the Actions:
";

const OBSERVER_RESPONSE_FORMAT: &str = r"
# Format of your response

You must use the following format for your response. Comments are in bold and should be removed from your response.
====================
## Observations: 
**What do you know to be true? What do you you don't know? What are your sources? Note down important information for later.**
- <...>
====================
";

const ORIENTER_RESPONSE_FORMAT: &str = r"
# Format of your response

You must use the following format for your response. Comments are in bold and should be removed from your response.
====================
## Orientation: 
**Plan the intermediate objectives to answer complete the original task. Maintain a list of current objectives updated as you go.**
- <...>
```
====================
";

const DECIDER_RESPONSE_FORMAT: &str = r"
# Format of your response

You must use the following format for your response. Comments are in bold and should be removed from your response.
====================
## Decision: 
**Decide what to do first to answer the question. Why? How will you if it succeeds? How will you if it fails?**
- <...>
====================
";

const ACTOR_RESPONSE_FORMAT: &str = r"
# Format of your response

You must use the following format for your response. Comments are in bold and should be removed from your response.
====================
## The ONLY Action: 
**Take a single Action consisting of exactly one pair of `tool_name` and `parameters`. Never give more than one YAML. **
```yaml
tool_name: <ToolName>
parameters:
    <...>  
```
We will take further action based on the response.
====================

Notes: 
- Action has the following fields: `tool_name` and `parameters` ONLY.
- `parameters` uses the format specified for the Tool.
- One Action at a time. No more. No less.
";

const OBSERVER_PROTO_INITIAL_RESPONSE: &str = r#"
## Observations:
- The given list to sort is [2, 3, 1, 4, 5].
- I need to sort this list in ascending order.
"#;

const OBSERVER_PROTO_SECOND_INPUT: &str = r#"
## Orientation:
- SandboxedPython can be used to sort the list.
- I need to provide only the `tool_name` and `parameters` fields for the SandboxedPython Tool.
- I expect the response of the Action to contains the field `stdout` with the sorted list and `stderr` empty.
- I need to use the Conclude Tool to terminate the task when I have the sorted list in plain text.
## Decision:
- We can use the sorted() function of Python to sort the list.
## The ONLY Action:
```yaml
tool_name: SandboxedPython
parameters:
  code: |
    lst = [2, 3, 1, 4, 5]
    sorted_list = sorted(lst)
    print(f"The sorted list is {sorted_list}")
```
We will take further action based on the response.
# Action SandboxedPython response:
```yaml
stdout: |
  The sorted list is [1, 2, 3, 4, 5]
stderr: ''
```
"#;

const OBSERVER_PROTO_SECOND_RESPONSE: &str = r"
## Observations:
- We needed to sort the list in ascending order.
- We have the response of the Action.
- We have the sorted list: [1, 2, 3, 4, 5].
";

const ORIENTER_PROTO_INITIAL_RESPONSE: &str = r#"
## Orientation:
- SandboxedPython can be used to sort the list.
- I need to provide only the `tool_name` and `parameters` fields for the SandboxedPython Tool.
- I expect the response of the Action to contains the field `stdout` with the sorted list and `stderr` empty.
- I need to use the Conclude Tool to terminate the task when I have the sorted list in plain text.
"#;

const ORIENTER_PROTO_SECOND_INPUT: &str = r#"
## Decision:
- We can use the sorted() function of Python to sort the list.
## The ONLY Action:
```yaml
tool_name: SandboxedPython
parameters:
  code: |
    lst = [2, 3, 1, 4, 5]
    sorted_list = sorted(lst)
    print(f"The sorted list is {sorted_list}")
```
We will take further action based on the response.
# Action SandboxedPython response:
```yaml
stdout: |
  The sorted list is [1, 2, 3, 4, 5]
stderr: ''
```
## Observations:
- We needed to sort the list in ascending order.
- We have the response of the Action.
- We have the sorted list: [1, 2, 3, 4, 5].
"#;

const ORIENTER_PROTO_SECOND_RESPONSE: &str = r"
## Orientation:
- I know the answer to the original question.
- I need to provide the `tool_name` and `parameters` fields for the Conclude Tool.
";

const DECIDER_PROTO_INITIAL_RESPONSE: &str = r#"
## Decision:
- We can use the sorted() function of Python to sort the list.
"#;

const DECIDER_PROTO_SECOND_INPUT: &str = r#"
## The ONLY Action:
```yaml
tool_name: SandboxedPython
parameters:
  code: |
    lst = [2, 3, 1, 4, 5]
    sorted_list = sorted(lst)
    print(f"The sorted list is {sorted_list}")
```
We will take further action based on the response.
# Action SandboxedPython response:
```yaml
stdout: |
  The sorted list is [1, 2, 3, 4, 5]
stderr: ''
```
## Observations:
- We needed to sort the list in ascending order.
- We have the response of the Action.
- We have the sorted list: [1, 2, 3, 4, 5].
## Orientation:
- I know the answer to the original question.
- I need to provide the `tool_name` and `parameters` fields for the Conclude Tool.
"#;

const DECIDER_PROTO_SECOND_RESPONSE: &str = r"
## Decision:
- Use the Conclude Tool to terminate the task with the sorted list.
";

const ACTOR_PROTO_INITIAL_RESPONSE: &str = r#"
## The ONLY Action:
```yaml
tool_name: SandboxedPython
parameters:
  code: |
    lst = [2, 3, 1, 4, 5]
    sorted_list = sorted(lst)
    print(f"The sorted list is {sorted_list}")
```
That's it for now. We will take further action based on the response.
"#;

const ACTOR_PROTO_SECOND_INPUT: &str = r#"
# Action SandboxedPython response:
```yaml
stdout: |
  The sorted list is [1, 2, 3, 4, 5]
stderr: ''
```
## Observations:
- We needed to sort the list in ascending order.
- We have the response of the Action.
- We have the sorted list: [1, 2, 3, 4, 5].
## Orientation:
- I know the answer to the original question.
- I need to provide the `tool_name` and `parameters` fields for the Conclude Tool.
## Decision:
- Use the Conclude Tool to terminate the task with the sorted list.
"#;

const ACTOR_PROTO_SECOND_RESPONSE: &str = r"
## The ONLY Action:
```yaml
tool_name: Conclude
parameters:
  original_question: |
    Sort in ascending order: [2, 3, 1, 4, 5]
  conclusion: |
    The ascending sorted list is [1, 2, 3, 4, 5].
```
";

enum AgentRole {
    Observer { prompt_manager: prompt::Manager },
    Orienter { prompt_manager: prompt::Manager },
    Decider { prompt_manager: prompt::Manager },
    Actor { prompt_manager: prompt::Manager },
}

impl Debug for AgentRole {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            AgentRole::Observer { .. } => write!(f, "Observer"),
            AgentRole::Orienter { .. } => write!(f, "Orienter"),
            AgentRole::Decider { .. } => write!(f, "Decider"),
            AgentRole::Actor { .. } => write!(f, "Actor"),
        }
    }
}

impl AgentRole {
    async fn convert_context_to_chat_history(
        &self,
        mut chat_history: ChatHistory,
        context: &Context,
    ) -> Result<ChatHistory, Error> {
        // build the examples
        let examples = self.build_examples();

        let prompt_manager = match self {
            AgentRole::Observer { prompt_manager } => prompt_manager,
            AgentRole::Orienter { prompt_manager } => prompt_manager,
            AgentRole::Decider { prompt_manager } => prompt_manager,
            AgentRole::Actor { prompt_manager } => prompt_manager,
        };

        // Add the prompts to the chat history
        prompt_manager
            .populate_chat_history(&mut chat_history, examples)
            .await;

        // Convert the context to a chat history
        // - get the latest 'Task' from the context
        let task = context.get_latest_task().unwrap();

        let task = prompt_manager.build_task_prompt(&task);

        // build the chat history from the context:
        // - group together Orientation, Decision, Action, ActionResult messages as a
        //   single chat entry from the User
        // - Observation messages become individual chat entries from the Assistant
        let mut user_msg = vec![];
        match self {
            AgentRole::Observer { .. } => {
                for m in &context.messages {
                    match m {
                        Message::Observation { content, .. } => {
                            if !user_msg.is_empty() {
                                // Add the user message to the chat history as a message from the
                                // User
                                chat_history
                                    .add_chitchat(ChatEntry {
                                        msg: user_msg.join("\n"),
                                        role: Role::User,
                                    })
                                    .await;

                                user_msg.clear();
                            }

                            // Add the observation to the chat history as a message from the
                            // Observer
                            chat_history
                                .add_chitchat(ChatEntry {
                                    msg: content.to_string(),
                                    role: Role::Assistant,
                                })
                                .await;
                        }
                        Message::Orientation { content, .. } => {
                            user_msg.push(content.clone());
                        }
                        Message::Decision { content, .. } => {
                            user_msg.push(content.clone());
                        }
                        Message::Action { content, .. } => {
                            user_msg.push(content.clone());
                        }
                        Message::ActionResult {
                            invocation_count,
                            tool_name,
                            outcome,
                            ..
                        } => {
                            let entry = format_outcome(&task, invocation_count, tool_name, outcome);

                            user_msg.push(entry);
                        }
                        _ => {
                            // Nothing
                        }
                    }
                }
            }

            AgentRole::Orienter { .. } => {
                for m in &context.messages {
                    match m {
                        Message::Observation { content, .. } => {
                            user_msg.push(content.clone());
                        }
                        Message::Orientation { content, .. } => {
                            if !user_msg.is_empty() {
                                // Add the user message to the chat history as a message from the
                                // User
                                chat_history
                                    .add_chitchat(ChatEntry {
                                        msg: user_msg.join("\n"),
                                        role: Role::User,
                                    })
                                    .await;

                                user_msg.clear();
                            }

                            // Add the observation to the chat history as a message from the
                            // Observer
                            chat_history
                                .add_chitchat(ChatEntry {
                                    msg: content.to_string(),
                                    role: Role::Assistant,
                                })
                                .await;
                        }
                        Message::Decision { content, .. } => {
                            user_msg.push(content.clone());
                        }
                        Message::Action { content, .. } => {
                            user_msg.push(content.clone());
                        }
                        Message::ActionResult {
                            invocation_count,
                            tool_name,
                            outcome,
                            ..
                        } => {
                            let entry = format_outcome(&task, invocation_count, tool_name, outcome);

                            user_msg.push(entry);
                        }
                        _ => {
                            // Nothing
                        }
                    }
                }
            }
            AgentRole::Decider { .. } => {
                for m in &context.messages {
                    match m {
                        Message::Observation { content, .. } => {
                            user_msg.push(content.clone());
                        }
                        Message::Orientation { content, .. } => {
                            user_msg.push(content.clone());
                        }
                        Message::Decision { content, .. } => {
                            if !user_msg.is_empty() {
                                // Add the user message to the chat history as a message from the
                                // User
                                chat_history
                                    .add_chitchat(ChatEntry {
                                        msg: user_msg.join("\n"),
                                        role: Role::User,
                                    })
                                    .await;

                                user_msg.clear();
                            }

                            chat_history
                                .add_chitchat(ChatEntry {
                                    msg: content.to_string(),
                                    role: Role::Assistant,
                                })
                                .await;
                        }
                        Message::Action { content, .. } => {
                            user_msg.push(content.clone());
                        }
                        Message::ActionResult {
                            invocation_count,
                            tool_name,
                            outcome,
                            ..
                        } => {
                            let entry = format_outcome(&task, invocation_count, tool_name, outcome);

                            user_msg.push(entry);
                        }
                        _ => {
                            // Nothing
                        }
                    }
                }
            }
            AgentRole::Actor { .. } => {
                for m in &context.messages {
                    match m {
                        Message::Observation { content, .. } => {
                            user_msg.push(content.clone());
                        }
                        Message::Orientation { content, .. } => {
                            user_msg.push(content.clone());
                        }
                        Message::Decision { content, .. } => {
                            user_msg.push(content.clone());
                        }
                        Message::Action { content, .. } => {
                            if !user_msg.is_empty() {
                                // Add the user message to the chat history as a message from the
                                // User
                                chat_history
                                    .add_chitchat(ChatEntry {
                                        msg: user_msg.join("\n"),
                                        role: Role::User,
                                    })
                                    .await;

                                user_msg.clear();
                            }

                            chat_history
                                .add_chitchat(ChatEntry {
                                    msg: content.to_string(),
                                    role: Role::Assistant,
                                })
                                .await;
                        }
                        Message::ActionResult {
                            invocation_count,
                            tool_name,
                            outcome,
                            ..
                        } => {
                            let entry = format_outcome(&task, invocation_count, tool_name, outcome);

                            user_msg.push(entry);
                        }
                        _ => {
                            // Nothing
                        }
                    }
                }
            }
        }

        if !user_msg.is_empty() {
            // Add the user message to the chat history as a message from the User
            chat_history
                .add_chitchat(ChatEntry {
                    msg: user_msg.join("\n"),
                    role: Role::User,
                })
                .await;
        }

        if chat_history.is_chitchat_empty() {
            // Add the recurring prompts to the chat history
            chat_history
                .add_chitchat(ChatEntry {
                    msg: task.to_prompt(),
                    role: Role::User,
                })
                .await;
        }

        // prune the history if needed
        chat_history.purge().await?;

        Ok(chat_history)
    }

    fn build_examples(&self) -> Vec<(String, String)> {
        match self {
            AgentRole::Observer { prompt_manager } => {
                let warmup_task =
                    prompt_manager.build_task_prompt("Sort in ascending order: [2, 3, 1, 4, 5]");

                vec![
                    (
                        warmup_task.to_prompt(),
                        OBSERVER_PROTO_INITIAL_RESPONSE.trim().to_string(),
                    ),
                    (
                        (format!(
                            "{}{}",
                            OBSERVER_PROTO_SECOND_INPUT.trim(),
                            warmup_task.to_prompt()
                        ))
                        .trim()
                        .to_string(),
                        OBSERVER_PROTO_SECOND_RESPONSE.trim().to_string(),
                    ),
                ]
            }
            AgentRole::Orienter { prompt_manager } => {
                let warmup_task =
                    prompt_manager.build_task_prompt("Sort in ascending order: [2, 3, 1, 4, 5]");

                vec![
                    (
                        warmup_task.to_prompt(),
                        ORIENTER_PROTO_INITIAL_RESPONSE.trim().to_string(),
                    ),
                    (
                        (format!(
                            "{}{}",
                            ORIENTER_PROTO_SECOND_INPUT.trim(),
                            warmup_task.to_prompt()
                        ))
                        .trim()
                        .to_string(),
                        ORIENTER_PROTO_SECOND_RESPONSE.trim().to_string(),
                    ),
                ]
            }
            AgentRole::Decider { prompt_manager } => {
                let warmup_task =
                    prompt_manager.build_task_prompt("Sort in ascending order: [2, 3, 1, 4, 5]");

                vec![
                    (
                        warmup_task.to_prompt(),
                        DECIDER_PROTO_INITIAL_RESPONSE.trim().to_string(),
                    ),
                    (
                        (format!(
                            "{}{}",
                            DECIDER_PROTO_SECOND_INPUT.trim(),
                            warmup_task.to_prompt()
                        ))
                        .trim()
                        .to_string(),
                        DECIDER_PROTO_SECOND_RESPONSE.trim().to_string(),
                    ),
                ]
            }
            AgentRole::Actor { prompt_manager } => {
                let warmup_task =
                    prompt_manager.build_task_prompt("Sort in ascending order: [2, 3, 1, 4, 5]");

                vec![
                    (
                        warmup_task.to_prompt(),
                        ACTOR_PROTO_INITIAL_RESPONSE.trim().to_string(),
                    ),
                    (
                        (format!(
                            "{}{}",
                            ACTOR_PROTO_SECOND_INPUT.trim(),
                            warmup_task.to_prompt()
                        ))
                        .trim()
                        .to_string(),
                        ACTOR_PROTO_SECOND_RESPONSE.trim().to_string(),
                    ),
                ]
            }
        }
    }
}

/// An agent
pub struct Agent {
    role: AgentRole,
    config: SapiensConfig,
    observer: WeakRuntimeObserver,
}

impl Agent {
    /// Create a new [`Agent`] with the role of an observer.
    pub async fn new_observer(
        config: SapiensConfig,
        toolbox: Toolbox,
        observer: WeakRuntimeObserver,
    ) -> Self {
        let system_prompt =
            "You are part of Sapiens agents and your role is to observe and report.".to_string();

        let prompt = "What are your observations?".to_string();

        let prompt_manager = prompt::Manager::new(
            toolbox,
            system_prompt,
            prompt,
            PREFIX.to_string(),
            TOOL_PREFIX.to_string(),
            OBSERVER_RESPONSE_FORMAT.to_string(),
        );

        Self {
            role: AgentRole::Observer { prompt_manager },
            config,
            observer,
        }
    }

    /// Create a new [`Agent`] with the role of an orienter.
    pub async fn new_orienter(
        config: SapiensConfig,
        toolbox: Toolbox,
        observer: WeakRuntimeObserver,
    ) -> Self {
        let system_prompt =
            "You are part of Sapiens agents and your role is to orient the other agents based on the observations."
                .to_string();

        let prompt = "What is your orientation?".to_string();

        let prompt_manager = prompt::Manager::new(
            toolbox,
            system_prompt,
            prompt,
            PREFIX.to_string(),
            TOOL_PREFIX.to_string(),
            ORIENTER_RESPONSE_FORMAT.to_string(),
        );

        Self {
            role: AgentRole::Orienter { prompt_manager },
            config,
            observer,
        }
    }

    /// Create a new [`Agent`] with the role of a decider.
    pub async fn new_decider(
        config: SapiensConfig,
        toolbox: Toolbox,
        observer: WeakRuntimeObserver,
    ) -> Self {
        let system_prompt =
            "You are part of Sapiens agents and your role is to decide what need to be done based on the observations and guidance you got."
                .to_string();

        let prompt = "What is your decision?".to_string();

        let prompt_manager = prompt::Manager::new(
            toolbox,
            system_prompt,
            prompt,
            PREFIX.to_string(),
            TOOL_PREFIX.to_string(),
            DECIDER_RESPONSE_FORMAT.to_string(),
        );

        Self {
            role: AgentRole::Decider { prompt_manager },
            config,
            observer,
        }
    }

    /// Create a new [`Agent`] with the role of an actor.
    pub async fn new_actor(
        config: SapiensConfig,
        toolbox: Toolbox,
        observer: WeakRuntimeObserver,
    ) -> Self {
        let system_prompt =
            "You are part of Sapiens agents and your role is to act on the world as it has been decided."
                .to_string();

        let prompt = "What is your action?".to_string();

        let prompt_manager = prompt::Manager::new(
            toolbox,
            system_prompt,
            prompt,
            PREFIX.to_string(),
            TOOL_PREFIX.to_string(),
            ACTOR_RESPONSE_FORMAT.to_string(),
        );

        Self {
            role: AgentRole::Actor { prompt_manager },
            config,
            observer,
        }
    }

    async fn convert_context_to_chat_history(
        &self,
        context: &Context,
    ) -> Result<ChatHistory, Error> {
        let max_token = self.config.model.context_size().await;

        // Create a new chat history
        let chat_history = ChatHistory::new(self.config.clone(), max_token);
        self.role
            .convert_context_to_chat_history(chat_history, context)
            .await
    }
}

#[async_trait::async_trait]
impl chains::Agent for Agent {
    type Error = Error;

    async fn act(&self, context: &Context) -> Result<Message, Error> {
        let chat_history = self.convert_context_to_chat_history(context).await?;

        // Query the model
        let input = chat_history.make_input();

        debug!(
            min_tokens = self.config.min_tokens_for_completion,
            max_tokens = self.config.max_tokens,
            role = ?self.role,
            "Querying model with {} entries",
            input.chat.len()
        );

        trace!("Querying model:\n{:#?}", input);

        let res = self
            .config
            .model
            .query(input, self.config.max_tokens)
            .await?;

        trace!("Got model response:\n{:#?}", res);

        // Show the message from the assistant
        if let Some(observer) = self.observer.upgrade() {
            observer
                .lock()
                .await
                .on_model_update(res.clone().into())
                .await;
        }

        // Return the response as a message
        match self.role {
            AgentRole::Observer { .. } => Ok(Message::Observation {
                content: res.msg,
                usage: res.usage,
            }),
            AgentRole::Orienter { .. } => Ok(Message::Orientation {
                content: res.msg,
                usage: res.usage,
            }),
            AgentRole::Decider { .. } => Ok(Message::Decision {
                content: res.msg,
                usage: res.usage,
            }),
            AgentRole::Actor { .. } => Ok(Message::Action {
                content: res.msg,
                usage: res.usage,
            }),
        }
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;

    use indoc::indoc;
    use insta::assert_debug_snapshot;

    use super::*;
    use crate::chains::Outcome;
    use crate::void_observer;

    #[tokio::test]
    async fn observer_converts_context_to_chat_history() {
        let context = build_dummy_context();

        let toolbox = Toolbox::default();

        let observer = void_observer();
        let weak_observer = Arc::downgrade(&observer);
        let agent = Agent::new_observer(Default::default(), toolbox, weak_observer).await;

        let chat_history = agent.convert_context_to_chat_history(&context).await;

        assert_debug_snapshot!(chat_history);
    }

    #[tokio::test]
    async fn orienter_converts_context_to_chat_history() {
        let mut context = build_dummy_context();

        context.add_message(Message::Observation {
            content: indoc! {r#"
            ## Observations:
            - We needed to sort the list in ascending order.
            - We have the response of the Action.
            - We have the sorted list: [1, 2, 3, 4, 5].
            "#
            }
            .trim()
            .to_string(),
            usage: None,
        });

        let toolbox = Toolbox::default();

        let observer = void_observer();
        let weak_observer = Arc::downgrade(&observer);
        let agent = Agent::new_orienter(Default::default(), toolbox, weak_observer).await;

        let chat_history = agent.convert_context_to_chat_history(&context).await;

        assert_debug_snapshot!(chat_history);
    }

    #[toki