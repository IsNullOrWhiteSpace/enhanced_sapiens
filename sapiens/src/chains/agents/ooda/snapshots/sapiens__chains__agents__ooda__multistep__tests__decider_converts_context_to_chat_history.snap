---
source: sapiens/src/chains/agents/ooda/multistep.rs
expression: chat_history
---
Ok(
    ChatHistory {
        config: Config {
            max_steps: 10,
            chain_type: SingleStepOODA,
            min_tokens_for_completion: 256,
            max_tokens: None,
        },
        max_token: 4096,
        context: [
            [system]: You are part of Sapiens agents and your role is to decide what need to be done based on the observations and guidance you got.,
            [user]: You are part of a group of cooperating assistants named Sapiens. Use available tools to answer the question as best as you can.
            You will collectively proceed iteratively using an OODA loop. Don't overstep your role.
            
            - Action response will be provided. 
            - Never produce the response of an Action. 
            - Only use YAML for the Action.
            - The loop will repeated until you have the answer to the original question. 
            - No task is complete until the Conclude Tool is used to provide the answer. 
            
            # Format of your response
            
            You must use the following format for your response. Comments are in bold and should be removed from your response.
            ====================
            ## Decision: 
            **Decide what to do first to answer the question. Why? How will you if it succeeds? How will you if it fails?**
            - <...>
            ====================
            
            # The following are the ONLY Tools one can use for the Actions:
            [],
        ],
        examples: [
            (
                [user]: # Your turn
                Original question: Sort in ascending order: [2, 3, 1, 4, 5]
                What is your decision?,
                [assistant]: ## Decision:
                - We can use the sorted() function of Python to sort the list.,
            ),
            (
                [user]: ## The ONLY Action:
                ```yaml
                tool_name: SandboxedPython
                parameters:
                  code: |
                    lst = [2, 3, 1, 4, 5]
                    sorted